### PR TITLE
[DebugInfo] Return complete variable info from getVarInfo by default

### PR DESCRIPTION
**Pull request by Snowy1803**
getVarInfo() now always returns a variable with a location and scope.
To opt out of this change, getVarInfo(false) returns an incomplete variable.
This can be used to work around bugs, but should only really be used for printing.

The complete var info will also contain the type, except for debug_values, as its type depends on another instruction, which may be inconsistent if called mid-pass.

All locations in debug variables are now also stripped of flags, to avoid issues when comparing or hashing debug variables.

Variable types are now also consistently object types, as there are no address types in Swift.

This PR also fixes some issues that this change uncovered:
- DeadObjectElimination using the wrong location and scope (now loses 18% less variables)
- SILCombine existential opening was using the wrong scope (loses 0.02% less variables)
- SILCombine emitting wrong debug info for eliminated enums (loses 0.02% less variables)

Overall, with this PR, we are losing 0.13% less variables at the SIL level. At the DWARF level, we have 1.9% more coverage and 2.5% more variables.

### CODE REVIEW

**Comment by Snowy1803** [comment]
@swift-ci please test

**Comment by Snowy1803** [comment]
@swift-ci please test source compatibility

File: lib/SILOptimizer/Mandatory/MovedAsyncVarDebugInfoPropagator.cpp [hunk 1]
```diff
@@ -347,8 +347,10 @@ struct DebugInfoPropagator {
     // var with a count already and return that value. If we did not, we insert
     // with the new count before expanding the set (initializing the map with
     // the correct value).
+    auto debugVariable = debugVar.withoutDIExpr();
+    debugVariable.Type = {};
     auto iter = dbgVarToDbgVarIndexMap.insert(
-        {debugVar, dbgVarToDbgVarIndexMap.size()});
+        {debugVariable, dbgVarToDbgVarIndexMap.size()});
```
Comment (adrian-prantl) [inline lib/SILOptimizer/Mandatory/MovedAsyncVarDebugInfoPropagator.cpp:353]: Should this take fragments into account and treat them as separate keys?

File: lib/SILOptimizer/Mandatory/DebugInfoCanonicalizer.cpp [hunk 2]
```diff
@@ -221,6 +221,9 @@ bool DebugInfoCanonicalizer::process() {
           LLVM_DEBUG(llvm::dbgs() << "        Has no var info?! Skipping!\n");
           continue;
         }
+        // Strip things we don't need in the map.
+        debugInfo->DIExpr = {};
```
Comment (adrian-prantl) [inline lib/SILOptimizer/Mandatory/DebugInfoCanonicalizer.cpp:225]: Same comment as below :-)

File: include/swift/SIL/DebugUtils.h [hunk 3]
```diff
@@ -437,16 +437,16 @@ struct DebugVarCarryingInst : VarDeclCarryingInst {
 
   Kind getKind() const { return Kind(VarDeclCarryingInst::getKind()); }
 
-  std::optional<SILDebugVariable> getVarInfo() const {
+  std::optional<SILDebugVariable> getVarInfo(bool complete = true) const {
```
Comment (adrian-prantl) [inline include/swift/SIL/DebugUtils.h:440]: Can you document what the complete flag does and under what circumstanced we would not want the default value here?

**Comment by Snowy1803** [comment]
@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]

**Comment by Snowy1803** [comment]
@swift-ci please test

**Comment by Snowy1803** [comment]
@swift-ci please test

**Pull request merged**